contracts/Vault.sol

```diff
@@ -781,32 +792,86 @@ contract Vault is
 
         for (uint8 i; i < idsLen; ++i) {
             uint256 tokenId = _ids[i];
+            uint256 amount = deposits[tokenId].amount;
 
-            Deposit memory _deposit = deposits[tokenId];
-            uint256 lockedUntil = _deposit.lockedUntil;
-            address claimerId = _deposit.claimerId;
+            _unsponsorSingle(tokenId, amount);
 
-            address owner = _deposit.owner;
-            uint256 amount = _deposit.amount;
+            sponsorAmount += amount;
+        }
 
-            if (owner != msg.sender) revert VaultNotAllowed();
-            if (lockedUntil > block.timestamp) revert VaultAmountLocked();
-            if (claimerId != address(0)) revert VaultNotSponsor();
+        _decreaseTotalSponsoredAndTransfer(_to, sponsorAmount);
+    }
+
+    /**
+     * Withdraws the specified sponsored amounts @param _amounts for the deposits with the ids provided
+     * in @param _ids and sends it to @param _to.
+     *
+     * @param _to Address that will receive the funds.
+     * @param _ids Array with the ids of the deposits.
+     * @param _amounts Array with the amounts to withdraw.
+     */
+    function _partialUnsponsor(
+        address _to,
+        uint256[] calldata _ids,
+        uint256[] calldata _amounts
+    ) internal {
+        uint256 sponsorAmount;
+        uint256 idsLen = _ids.length;
+
+        for (uint8 i; i < idsLen; ++i) {
+            uint256 tokenId = _ids[i];
+            uint256 amount = _amounts[i];
+
+            _unsponsorSingle(tokenId, amount);
 
             sponsorAmount += amount;
+        }
 
-            delete deposits[tokenId];
+        _decreaseTotalSponsoredAndTransfer(_to, sponsorAmount);
+    }
+
+    /**
+     * Validates conditions for unsponsoring amount @param _amount of the deposit with the id @param _id.
+     *
+     * @notice The NFTs of the deposit will be burned if the deposited amount is equal to the amount being withdrawn.
+     *
+     * @param _id Id of the deposit.
+     * @param _amount Amount to be unsponsored/withdrawn.
+     */
+    function _unsponsorSingle(uint256 _id, uint256 _amount) internal {
+        Deposit memory _deposit = deposits[_id];
 
-            emit Unsponsored(tokenId);
+        if (_deposit.owner != msg.sender) revert VaultNotAllowed();
+        if (_deposit.lockedUntil > block.timestamp) revert VaultAmountLocked();
+        if (_deposit.claimerId != address(0)) revert VaultNotSponsor();
+        if (_deposit.amount < _amount)
+            revert VaultCannotWithdrawMoreThanAvailable();
+
+        if (_amount != _deposit.amount) {
```

Comment:
I think here it's cleaner to use if return; instead of if else because the else block is at the end of the function

Reply:
[changed](https://github.com/lindy-labs/sc_solidity-contracts/pull/188/commits/41ba2122bd097c74eb996f930d999acebe9f19f4)